crates/fixedpointmath/src/lib.rs

```diff
@@ -1,422 +1,406 @@
+                bail!(
+                    "FixedPoint operation overflowed: {} * {} / {}",
+                    self.to_scaled_string(),
+                    other.to_scaled_string(),
+                    divisor.to_scaled_string(),
+                )
+            })?)?,
+        )
+    }
+
+    pub fn mul_div_up(&self, other: Self, divisor: Self) -> Result<Self> {
+        if divisor.is_zero() {
+            bail!("cannot divide by zero");
+        }
+        let (u512, remainder) = self
+            .abs()
+            .to_u256()?
+            .full_mul(other.abs().to_u256()?.into())
+            .div_mod(divisor.abs().to_u256()?.into());
+        let rounded_u512 = u512 + (remainder.gt(&U512::zero()) as u128);
+        Self::new(
+            self.sign().flip_if(other.sign() != divisor.sign()),
+            T::from_u256(U256::try_from(rounded_u512).or_else(|_| {
+                bail!(
+                    "FixedPoint operation overflowed: {} * {} / {}",
+                    self.to_scaled_string(),
+                    other.to_scaled_string(),
+                    divisor.to_scaled_string()
+                )
+            })?)?,
+        )
     }
 
-    pub fn div_down(self, other: FixedPoint) -> FixedPoint {
-        self.mul_div_down(fixed!(1e18), other)
+    pub fn mul_down(&self, other: Self) -> Result<Self> {
+        self.mul_div_down(other, self.one())
     }
 
-    pub fn div_up(self, other: FixedPoint) -> FixedPoint {
-        self.mul_div_up(fixed!(1e18), other)
+    pub fn mul_up(&self, other: Self) -> Result<Self> {
+        self.mul_div_up(other, self.one())
     }
 
-    pub fn pow(self, y: FixedPoint) -> Result<FixedPoint> {
-        // If the exponent is 0, return 1.
-        if y == fixed!(0) {
-            return Ok(fixed!(1e18));
-        }
+    pub fn div_down(self, other: Self) -> Result<Self> {
+        self.mul_div_down(self.one(), other)
+    }
 
-        // If the base is 0, return 0.
-        if self == fixed!(0) {
-            return Ok(fixed!(0));
-        }
+    pub fn div_up(self, other: Self) -> Result<Self> {
+        self.mul_div_up(self.one(), other)
+    }
 
-        // Using properties of logarithms we calculate x^y:
-        // -> ln(x^y) = y * ln(x)
-        // -> e^(y * ln(x)) = x^y
-        let y_int256 = I256::try_from(y)?;
+    pub fn pow(self, y: Self) -> Result<Self> {
+        let one = self.one();
 
-        // Compute y*ln(x)
-        // Any overflow for x will be caught in _ln() in the initial bounds check
-        let lnx: I256 = Self::ln(I256::from_raw(self.0))?;
-        let mut ylnx: I256 = y_int256.wrapping_mul(lnx);
-        ylnx = ylnx.wrapping_div(int256!(1e18));
+        // If the exponent is negative, return 1 / x^abs(y).
+        if y.is_negative() {
+            let abs_result = self.pow(y.abs())?;
 
-        // Calculate exp(y * ln(x)) to get x^y
-        Self::exp(ylnx)?.try_into()
-    }
+            // FIXME: What's the correct return value here?
```

Comment:
python float handles this by including special values `inf` and `-inf` and `nan`. https://github.com/delvtech/fixedpointmath/blob/main/fixedpointmath/fixed_point.py#L164
then we check that member variable whenever people want to perform an operation.
You can see all of the rules here: https://github.com/delvtech/fixedpointmath/blob/main/tests/test_fp_nonfinite_checks.py